sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala

```diff
@@ -29,36 +29,45 @@ import org.apache.spark.sql.rapids.execution.TrampolineUtil
  */
 class RapidsHostMemoryStore(
     maxSize: Long,
+    pinnedSize: Long,
     catalog: RapidsBufferCatalog = RapidsBufferCatalog.singleton,
     deviceStorage: RapidsDeviceMemoryStore = RapidsBufferCatalog.getDeviceStorage)
     extends RapidsBufferStore(StorageTier.HOST, catalog) {
   private[this] val pool = HostMemoryBuffer.allocate(maxSize, false)
   private[this] val addressAllocator = new AddressSpaceAllocator(maxSize)
   private[this] var haveLoggedMaxExceeded = false
 
-  // Returns an allocated host buffer and whether the allocation is from the internal pool
-  private def allocateHostBuffer(size: Long): (HostMemoryBuffer, Boolean) = {
+  private object AllocationMode extends Enumeration {
+    type AllocationMode = Value
+    val Pinned, Pooled, Direct = Value
+  }
+  import AllocationMode._
+
+  // Returns an allocated host buffer and its allocation mode
+  private def allocateHostBuffer(size: Long): (HostMemoryBuffer, AllocationMode) = {
     // spill to keep within the targeted size
-    val amountSpilled = synchronousSpill(math.max(maxSize - size, 0))
+    val targetSize = math.max(maxSize + pinnedSize - size, 0)
+    val amountSpilled = synchronousSpill(targetSize)
     if (amountSpilled != 0) {
       logInfo(s"Spilled $amountSpilled bytes from the host memory store")
       TrampolineUtil.incTaskMetricsDiskBytesSpilled(amountSpilled)
     }
 
     var buffer: HostMemoryBuffer = null
     while (buffer == null) {
-      buffer = PinnedMemoryPool.tryAllocate(size)
-      if (buffer != null) {
-        return (buffer, false)
-      }
-
-      if (size > maxSize) {
+      if (size > maxSize && size > pinnedSize) {
```

Comment:
The pinned memory pool is essentially a "bonus" area for spilled host memory buffers -- it is not guaranteed to be there.  Therefore I think we should only be checking against `maxSize` here for warning purposes, and only when we are unable to allocate in the pinned memory pool.

Reply:
Done.